subsys/net/ip/net_tls.c

```diff
@@ -0,0 +1,735 @@
+
+	for (i = 0; i < ARRAY_SIZE(credentials); i++) {
+		if (credentials[i].type == NET_TLS_CREDENTIAL_UNUSED) {
+			return &credentials[i];
+		}
+	}
+
+	return NULL;
+}
+
+static struct net_tls_credential *credential_get(
+		sec_tag_t tag, enum net_tls_credential_type type)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(credentials); i++) {
+		if (credentials[i].type == type && credentials[i].tag == tag) {
+			return &credentials[i];
+		}
+	}
+
+	return NULL;
+}
+
+static struct net_tls_credential *credential_next_get(
+		sec_tag_t tag, struct net_tls_credential *iter)
+{
+	int i;
+
+	if (!iter) {
+		iter = credentials;
+	} else {
+		iter++;
+	}
+
+	for (i = iter - credentials; i < ARRAY_SIZE(credentials); i++) {
+		if (credentials[i].type != NET_TLS_CREDENTIAL_UNUSED &&
+		    credentials[i].tag == tag) {
+			return &credentials[i];
+		}
+	}
+
+	return NULL;
+}
+
+static int tls_tx(void *ctx, const unsigned char *buf, size_t len)
+{
+	struct net_context *context = ctx;
+	struct net_pkt *pkt;
+	int bytes, ret;
+
+	pkt = net_pkt_get_tx(context, TIMEOUT_TLS_TX_MS);
+
+	if (!pkt) {
+		return -EIO;
+	}
+
+	bytes = net_pkt_append(pkt, len, (u8_t *)buf, TIMEOUT_TLS_TX_MS);
+
+	ret = net_context_output(context, pkt, &context->remote);
+	if (ret < 0) {
+		net_pkt_unref(pkt);
+		return ret;
+	}
+
+	return bytes;
+}
+
+static int tls_rx(void *ctx, unsigned char *buf, size_t len)
+{
+	struct net_context *context = ctx;
+	int ret;
+	u16_t len_left;
+
+	if (!context->tls->rx_pkt) {
+		context->tls->rx_pkt =
+			k_fifo_get(&context->tls->rx_fifo, K_NO_WAIT);
+		if (context->tls->rx_pkt) {
+			context->tls->rx_offset =
+				net_pkt_appdata(context->tls->rx_pkt) -
+				context->tls->rx_pkt->frags->data;
+		} else {
+			return MBEDTLS_ERR_SSL_WANT_READ;
+		}
+	}
+
+	len_left = net_pkt_get_len(context->tls->rx_pkt) -
+		   context->tls->rx_offset;
+	ret = net_frag_linearize(buf, len, context->tls->rx_pkt,
+				 context->tls->rx_offset,
+				 MIN(len, len_left));
+	if (ret > 0) {
+		context->tls->rx_offset += ret;
+	}
+
+	if (context->tls->rx_offset >=
+	    net_pkt_get_len(context->tls->rx_pkt)) {
+		if (net_context_get_type(context) == SOCK_STREAM) {
+			net_context_update_recv_wnd(
+				context,
+				net_pkt_appdatalen(context->tls->rx_pkt));
+		}
+		net_pkt_unref(context->tls->rx_pkt);
+		context->tls->rx_pkt = NULL;
+		context->tls->rx_offset = 0;
+	}
+
+	return ret;
+}
+
+static void tls_recv_cb(struct net_context *context, struct net_pkt *pkt,
+			int status, void *user_data)
+{
+	struct net_pkt *decrypted_pkt;
+	int len, ret = 0;
+
+	if (!pkt) {
+		/* Forward EOF */
+		if (context->tls->tls_cb) {
+			context->tls->tls_cb(context, pkt, status, user_data);
+		}
+		return;
+	}
+
+	k_fifo_put(&context->tls->rx_fifo, pkt);
+
+	if (net_context_get_type(context) == SOCK_STREAM) {
+		net_context_update_recv_wnd(context, -net_pkt_appdatalen(pkt));
+	}
+
+	/* Process it as application data only after handshake is over,
+	 * otherwise packet will be consumed by handshake.
```

Comment:
Is this comment right? Would seem that the purpose of code is to avoid processing handshake packet as app data.